gapil/compiler/types.go

```diff
@@ -129,119 +131,132 @@ func (c *C) declareTypes() {
+				codegen.Field{Name: RefRefCount, Type: c.T.Uint32},
+				codegen.Field{Name: RefArena, Type: c.T.ArenaPtr},
+				codegen.Field{Name: RefValue, Type: c.T.Target(t.To)},
+			)
+		}
 	}
 
 	// Build all the map types.
 	c.defineMapTypes()
 
 	c.buildRefRels()
 
-	globalsFields := make([]codegen.Field, len(c.API.Globals))
-	for i, g := range c.API.Globals {
-		globalsFields[i] = codegen.Field{Name: g.Name(), Type: c.T.Target(g.Type)}
+	apiGlobals := make([]codegen.Field, len(c.APIs))
+	for i, api := range c.APIs {
+		fields := make([]codegen.Field, len(api.Globals))
+		for i, g := range api.Globals {
+			fields[i] = codegen.Field{Name: g.Name(), Type: c.T.Target(g.Type)}
+		}
+		apiGlobals[i] = codegen.Field{
+			Name: api.Name(),
+			Type: c.T.Struct(api.Name(), fields...),
+		}
 	}
-	c.T.Globals.SetBody(false, globalsFields...)
+	c.T.Globals.SetBody(false, apiGlobals...)
 
 	// Build storage types for the capture's memory layout.
 	c.T.CaptureTypes = c.StorageTypes(c.Settings.CaptureABI.MemoryLayout, "S_")
 
 	// Build conversion functions between target and capture types.
 	if c.Settings.CaptureABI != c.T.targetABI {
-		for _, t := range c.API.Classes {
-			if !semantic.IsStorageType(t) {
-				continue
-			}
-			captureTypePtr := c.T.Pointer(c.T.Capture(t))
-			targetTypePtr := c.T.Pointer(c.T.Target(t))
-
-			copyToTarget := c.M.Function(c.T.Void, "S_"+t.Name()+"_copy_to_target", c.T.CtxPtr, captureTypePtr, targetTypePtr).
-				LinkOnceODR().
-				Inline()
-
-			c.T.captureToTarget[t] = copyToTarget
-			c.Build(copyToTarget, func(s *S) {
-				src := s.Parameter(1).SetName("src")
-				dst := s.Parameter(2).SetName("dst")
-				for _, f := range t.Fields {
-					firstElem := src.Index(0, f.Name()).LoadUnaligned()
-					dst.Index(0, f.Name()).Store(c.castCaptureToTarget(s, f.Type, firstElem))
-				}
-			})
-
-			copyToCapture := c.M.Function(c.T.Void, "T_"+t.Name()+"_copy_to_capture", c.T.CtxPtr, targetTypePtr, captureTypePtr).
-				LinkOnceODR().
-				Inline()
-
-			c.T.targetToCapture[t] = copyToCapture
-			c.Build(copyToCapture, func(s *S) {
-				src := s.Parameter(1).SetName("src")
-				dst := s.Parameter(2).SetName("dst")
-				for _, f := range t.Fields {
-					firstElem := src.Index(0, f.Name()).Load()
-					dst.Index(0, f.Name()).StoreUnaligned(c.castTargetToCapture(s, f.Type, firstElem))
+		for _, api := range c.APIs {
+			for _, t := range api.Classes {
+				if !semantic.IsStorageType(t) {
+					continue
 				}
-			})
+				captureTypePtr := c.T.Pointer(c.T.Capture(t))
+				targetTypePtr := c.T.Pointer(c.T.Target(t))
+
+				copyToTarget := c.M.Function(c.T.Void, "S_"+t.Name()+"_copy_to_target", c.T.CtxPtr, captureTypePtr, targetTypePtr).
+					LinkOnceODR().
+					Inline()
+
+				c.T.captureToTarget[t] = copyToTarget
+				c.Build(copyToTarget, func(s *S) {
+					src := s.Parameter(1).SetName("src")
+					dst := s.Parameter(2).SetName("dst")
+					for _, f := range t.Fields {
+						firstElem := src.Index(0, f.Name()).LoadUnaligned()
```

Comment:
I don't love super deep nesting, but I don't actually have a solution for you right now anyway.

Reply:
Ack. I'll see if I can clean it up when I visit this code again.